### DISCUSSION TITLE
post_233883

### DISCUSSION THREAD

**Discussion by giscus** [cincuranet/blog-comments#1353] [Comments]
# post_233883

https://www.tabsoverspaces.com/233883-simple-websocket-client-and-server-application-using-dotnet

**Comment by dave-lewis73** [comment]
This is a great example of how easy it is to set up a simple websocket.  However, I have a use case where I need to send data from a C# application to a JavaScript web page using websockets.  How would I go about doing that in your example?  I need to to send a simple JSON string from the C# client that you showed how to create to the socket server and then that information is passed on to the JavaScript page.  It is a seemingly super simple concept, but I have been struggling with how to do it for weeks.  It is just uni-directional passing of information from one client that is in C# to a web page that is written in JavaScript.  Send to server and then pass along and receive in web page.

**Reply by cincuranet** [reply]
Um, same way. You'll have WS server using C# and open connection to it from web page using JS (or maybe even through stuff like WASM/Blazor). Once that connection is established, whenever server needs, it will use the socket to send some data through it (`webSocket.SendAsync` in my example).

**Reply by dave-lewis73** [reply]
I have tried as you suggested, but it is not working for me.  I am sending my JSON string from the C# client to the server, and it is being received properly.  I am then sending a response, which is simply returning the JSON back.  All the while that this is happening, I have the JavaScript page showing as connected to the server, but it does not receive the JSON that has been sent on from the server.  The C# client application that sent the message in the first place receives the response back, as I am able to write it to the console.  However, the JavaScript doesn't receive anything.  I get nothing through "onmessage" in the JavaScript.  I am clearly doing something wrong because when I comment out all of my code and return things back to your base code, I am able to get both the C# client application and the JavaScript page receiving the time stamp information that your server sends.  Why would the JavaScipt page not receive the forwarded JSON string from the C# client, if it can receive the time stamp?  Can the client that sends the information initially be the only one that receives a return?  If so, how do I get the JavaScript file to receive the JSON?  Please let me knowif you any further information.  I appreciate any help you can provide.  

The following is the server code that receives the JSON initially and then sends it back:

using System.Net;
using System.Net.WebSockets;
using System.Text;

Console.Title = "Server";

var builder = WebApplication.CreateBuilder(args);
builder.WebHost.UseUrls("http://localhost:8080");
var app = builder.Build();

app.UseWebSockets();
app.Map("/ws", async context =>
{
    if (context.WebSockets.IsWebSocketRequest)
    {
        using var ws = await context.WebSockets.AcceptWebSocketAsync();
        {
            while (true)
            {
                var result = await ws.ReceiveAsync(buf, CancellationToken.None);
                if (result.MessageType == WebSocketMessageType.Close)
                {
                    await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, null, CancellationToken.None);
                    Console.WriteLine(result.CloseStatusDescription);
                }
                else
                {
                    string receiveStr = Encoding.ASCII.GetString(buf, 0, result.Count);
                    Console.WriteLine("Received:  " + receiveStr);

                    byte[] data = Encoding.ASCII.GetBytes(receiveStr);
                    await ws.SendAsync(data, WebSocketMessageType.Text, true, CancellationToken.None);
                    await Task.Delay(1000);
                }
            }
        }
    }
    else
    {
        context.Response.StatusCode = (int)HttpStatusCode.BadRequest;
    }
});
await app.RunAsync();

**Reply by cincuranet** [reply]
This `Can the client that sends the information initially be the only one that receives a return?` confuses me a little bit. Are you trying to do some kind of broadcast/multicast? Because WS is regular "point-to-point" type of connection. If you want "broadcast/multicast", you need to create it on top of that, or maybe rather use i.e. SignalR.

**Comment by LajpatMalhi** [comment]
i want to create Websocket in .NET
